front/text/plain/convert.go

```diff
@@ -132,11 +132,12 @@ func ToHTML(text string) string {
 		return text
 	}
 
+	text = strings.ReplaceAll(text, "\n\n", "</p><p>")
+	text = strings.ReplaceAll(text, "\n", "<br/>")
+
 	var b strings.Builder
-	for _, line := range lines {
```

Comment:
How about this instead?

```
	for _, line := range lines {
		if line == "" {
			w.WriteString("<br/>")
			continue
                }
 		b.WriteString("<p>")
		b.WriteString(line)
		b.WriteString("</p>")
	}
```

Each non-empty line is wrapped with <p></p>, otherwise replaced with <br/>. I took what seems to be a representative sample of posts from various servers, and this should mimic what the vast majority do. Don't forget that the input that passes through this function is plain text, not gemtext.

Reply:
Ah, no! Here's why:

So before, this input

    Lorem ipsum dolor
    sit amet

    consectetur adipisicing elit,
    sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua

would turn into

    Lorem ipsum dolor

    sit amet

    consectetur adipisicing elit,

    sed do eiusmod tempor incididunt ut

    labore et dolore magna aliqua

(And that's *after* the whitespace collapsing that Mastodon does. Without whitespace collapsing, there are even more blank lines in there.)

With my patch, the input displays the same on Mastodon as it does on the tootik side:


    Lorem ipsum dolor
    sit amet

    consectetur adipisicing elit,
    sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua

With your change, it would instead look like this

	Lorem ipsum dolor

	sit amet


	consectetur adipisicing elit,

	sed do eiusmod tempor incididunt ut

	labore et dolore magna aliqua


The explicit BR overrides the whitespace collapse making it even more weird-looking.

An argument could be made that it should become


    Lorem ipsum dolor sit amet

    consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

Since people are using this who are used to the way Gemini line wrapping works, I would argue against that, but a stronger case could be made for that than for inserting extra blank lines. Which Mastodon server don't do either.

Reply:
I see. This is more consistent with Mastodon at the cost of slightly weird posts if you're used gemtext (where most lines are actually paragraphs).

Many posts in Mastodon look like this:

```
line 1
line 2

line 3
```

But this is uncommon in gemtext, which is more likely to look like this:

```
line 1

line 2

line 3
```

If I have to choose between the two, it must be the first option. I'd add more tests, though, for weird cases like posts that begin with empty lines, or have nothing but empty lines.

Reply:
What I typically see on Mastodon and Fedi is the same as I see in gemtext:


    This is one paragraph.

    This is a second paragraph.

    Roses are red,
    violets are blue,
    the third paragraph's a poem
    through and through.

    This is the fourth paragraph.


As long as there was no brs (no poems), like this:

    This is one paragraph

    This is another paragraph

    This is a third, non-poetic paragraph,

    This is the fourth paragraph


The old FromHTML was preserving that on the gemside exactly:

    This is one paragraph

    This is another paragraph

    This is a third, non-poetic paragraph,

    This is the fourth paragraph


But with BRs, it would treat them as paragraphs, so the old version of FromHTML would do this to the poem version:


    This is one paragraph.

    This is a second paragraph.

    Roses are red,

    violets are blue,

    the third paragraph's a poem

    through and through.

    This is the fourth paragraph.


(And since Akkoma has a bug that uses br br instead of p, and I'm not defending Akkoma on that, but that made the poem-splitting behavior more visible.)


Now, let's consider the same original roses-poem as if it were input from the gem side.
The intent of my commit here is to turn that into


    <p>This is one paragraph.</p>
    <p>This is a second paragraph.</p>

    <p>Roses are red,</br>
    violets are blue,</br>
    the third paragraph's a poem</br>
    through and through.</p>

    <p>This is the fourth paragraph.</p>

Reply:
Maybe you're saying that


    This is one paragraph.

    This is a second paragraph.

    Roses are red,
    violets are blue,
    the third paragraph's a poem
    through and through.

    This is the fourth paragraph.

should turn into:

    This is one paragraph.

    This is a second paragraph.

    Roses are red, violets are blue, the third paragraph's a poem through and through.

    This is the fourth paragraph.

I think it's reasonable to argue that **FromHTML** should do that. Maybe. I think it shouldn't, but a case could be made. But I don't think **ToHTML** should do that. Gemini users are used to their newlines being hard.

Reply:
> Gemini users are used to their newlines being hard.

Yes, and Mastodon users expect the opposite, but doing what looks more like gemtext was a mistake and your PR can fix this. I'll write tests on top of this branch, based on the FromHTML() tests - if this PR's ToHTML() can "reverse" what it does, it's perfect.

Reply:
Nice! I modded some of the tests already.